File: Plugins/asm-transform/src/main/java/com/mrcd/transform/concurrent/Worker.java
package com.mrcd.transform.concurrent;

import java.io.IOException;
import java.util.LinkedList;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Future;

/**
 * Created by im_dsd on 2021/3/26
 */
public class Worker {
    protected final LinkedList<Future<?>> futures = new LinkedList<>();
    protected ExecutorService executor;

    Worker(ExecutorService executor) {
        this.executor = executor;
    }

    public void execute(Runnable runnable) {
        futures.add(executor.submit(runnable));
    }

    public <T> Future<T> submit(Callable<T> callable) {
        Future<T> future = executor.submit(callable);
        futures.add(future);
        return future;
    }

    public void await() throws IOException {
        Future<?> future;
        while ((future = futures.pollFirst()) != null) {
            try {
                future.get();
            } catch (ExecutionException | InterruptedException e) {
                if (e.getCause() instanceof IOException) {
                    throw (IOException) e.getCause();
                } else if (e.getCause() instanceof RuntimeException) {
                    throw (RuntimeException) e.getCause();
                } else if (e.getCause() instanceof Error) {
                    throw (Error) e.getCause();
                }
                throw new RuntimeException(e.getCause());
            }
        }
    }
}

File: Plugins/base-asm-transform/src/main/java/com/dsd/mrcd/transform/concurrent/Worker.kt
package com.dsd.mrcd.transform.concurrent

import java.io.IOException
import java.util.*
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future

class Worker internal constructor(var executor: ExecutorService) {
    private val futures: LinkedList<Future<*>> = LinkedList<Future<*>>()

    fun execute(runnable: Runnable?) {
        runnable ?: return
        futures.add(executor.submit(runnable))
    }

    fun <T> submit(callable: Callable<T>?): Future<T> {
        val future = executor.submit(callable)
        futures.add(future)
        return future
    }

    @Throws(IOException::class)
    fun await() {
        var future: Future<*>
        while (futures.pollFirst().also { future = it } != null) {
            try {
                future.get()
            } catch (e: ExecutionException) {
                when (e.cause) {
                    is IOException -> {
                        throw (e.cause as IOException)
                    }
                    is RuntimeException -> {
                        throw (e.cause as RuntimeException)
                    }
                    is Error -> {
                        throw (e.cause as Error)
                    }
                    else -> throw RuntimeException(e.cause)
                }
            } catch (e: InterruptedException) {
                when (e.cause) {
                    is IOException -> {
                        throw (e.cause as IOException)
                    }
                    is RuntimeException -> {
                        throw (e.cause as RuntimeException)
                    }
                    is Error -> {
                        throw (e.cause as Error)
                    }
                    else -> throw RuntimeException(e.cause)
                }
            }
        }
    }
}